File: services/purchase-service/models/pr.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type PRStatus string

const (
	PRStatusDraft    PRStatus = "DRAFT"
	PRStatusPending  PRStatus = "PENDING"
	PRStatusApproved PRStatus = "APPROVED"
	PRStatusRejected PRStatus = "REJECTED"
)

type PurchaseRequest struct {
	gorm.Model
	ID          uint     `gorm:"primaryKey"`
	PRNumber    string   `gorm:"unique;not null"`
	RequesterID uint     `gorm:"not null"`
	Department  string   `gorm:"not null"`
	Purpose     string   `gorm:"type:text"` // Purpose of the purchase request
	Status      PRStatus `gorm:"type:varchar(20);default:'DRAFT'"`
	WorkflowID  string   `gorm:"type:varchar(100)"`
	Items       []PRItem `gorm:"foreignKey:PRID"`
	IsDeleted   bool     `gorm:"default:false"`
	CreatedAt   time.Time
	UpdatedAt   time.Time
	DeletedAt   gorm.DeletedAt `gorm:"index"`
}

type PRItem struct {
	gorm.Model
	ID                   uint   `gorm:"primaryKey"`
	PRID                 uint   `gorm:"not null;index"`
	SKU                  string `gorm:"not null;index"` // Stock Keeping Unit
	ItemName             string `gorm:"not null"`
	Description          string
	Quantity             int     `gorm:"not null"`
	PricePerUnit         float64 `gorm:"type:decimal(15,2)"`
	Discount             float64 `gorm:"type:decimal(15,2)"`
	DiscountUnit         string  `gorm:"type:varchar(10)"` // % or BAHT
	TotalPrice           float64 `gorm:"type:decimal(15,2)"`
	RequiredDate         time.Time
	CreatedAt            time.Time
	UpdatedAt            time.Time
}

File: client/src/lib/api/pr.ts
import { request } from "./client";

export interface PRItem {
    ID: number;
    PRID: number;
    CreatedAt: string;
    UpdatedAt: string;
    SKU: string;
    ItemName: string;
    Description: string;
    Quantity: number;
    PricePerUnit: number;
    Discount: number;
    DiscountUnit: string;
    TotalPrice: number;
    RequiredDate: string; // Optional flag to indicate if the item is marked for deletion
}

export interface PurchaseRequest {
    ID: number;
    CreatedAt: string;
    UpdatedAt: string;
    DeletedAt: string | null;
    PRNumber: string;
    Purpose: string;
    Department: string;
    Status: string;
    RequesterID: number;
    WorkflowID: string;
    IsDeleted: boolean;
    Items: PRItem[];
}

export const prApi = {
    getPurchaseRequests: () => request<PurchaseRequest[]>("/purchase/pr"),

    getPurchaseRequestById: (id: number) => request<PurchaseRequest>(`/purchase/pr/${id}`),

    createPurchaseRequest: (data: {
        department: string;
        purpose: string;
        items: Array<{
            sku: string;
            item_name: string;
            description: string;
            quantity: number;
            price_per_unit: number;
            discount?: number;
            discount_unit?: string;
            required_date: string;
        }>;
    }) =>
        request<{ message: string; data: PurchaseRequest }>("/purchase/pr", {
            method: "POST",
            body: JSON.stringify(data),
        }),

    editPurchaseRequest: (id: number, data: {
        department: string;
        purpose: string;
        items: Array<{
            sku: string;
            item_name: string;
            description: string;
            quantity: number;
            price_per_unit: number;
            discount?: number;
            discount_unit?: string;
            required_date: string;
        }>;
    }) =>
        request<{ message: string; data: PurchaseRequest }>(`/purchase/pr/${id}`, {
            method: "PUT",
            body: JSON.stringify(data),
        }),

    submitPurchaseRequest: (id: number) =>
        request<{ message: string; data: PurchaseRequest }>(`/purchase/pr/${id}/submit`, {
            method: "POST",
        }),

    deletePurchaseRequest: (id: number) =>
        request<{ message: string }>(`/purchase/pr/${id}`, {
            method: "DELETE",
        }),

};
